core/base/src/TROOT.cxx

```diff
@@ -2574,14 +2544,258 @@ const char**& TROOT::GetExtraInterpreterArgs() {
+/// Get the macro directory in the installation. Static utility function.
+
+const TString& TROOT::GetMacroDir() {
+#ifdef ROOTMACRODIR
+   if (IgnorePrefix()) {
+#endif
+      static TString rootmacrodir;
+      if (rootmacrodir.IsNull()) {
+         rootmacrodir = "macros";
+         gSystem->PrependPathName(GetRootSys(), rootmacrodir);
+      }
+      return rootmacrodir;
+#ifdef ROOTMACRODIR
+   } else {
+      const static TString rootmacrodir = ROOTMACRODIR;
+      return rootmacrodir;
+   }
+#endif
+}
+
+////////////////////////////////////////////////////////////////////////////////
 /// Get the tutorials directory in the installation. Static utility function.
 
-const char *TROOT::GetTutorialsDir()
-{
+const TString& TROOT::GetTutorialsDir() {
 #ifdef ROOTTUTDIR
-   return ROOTTUTDIR;
-#else
-   static TString tutdir = TString(gRootDir) + "/tutorials";
-   return tutdir;
+   if (IgnorePrefix()) {
+#endif
+      static TString roottutdir;
+      if (roottutdir.IsNull()) {
+         roottutdir = "tutorials";
+         gSystem->PrependPathName(GetRootSys(), roottutdir);
+      }
+      return roottutdir;
+#ifdef ROOTTUTDIR
+   } else {
+      const static TString roottutdir = ROOTTUTDIR;
+      return roottutdir;
+   }
+#endif
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Get the source directory in the installation. Static utility function.
+
+const TString& TROOT::GetSourceDir() {
+#ifdef ROOTSRCDIR
+   if (IgnorePrefix()) {
+#endif
+      static TString rootsrcdir;
+      if (rootsrcdir.IsNull()) {
+         rootsrcdir = "src";
+         gSystem->PrependPathName(GetRootSys(), rootsrcdir);
+      }
+      return rootsrcdir;
+#ifdef ROOTSRCDIR
+   } else {
+      const static TString rootsrcdir = ROOTSRCDIR;
+      return rootsrcdir;
+   }
+#endif
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Get the icon path in the installation. Static utility function.
+
+const TString& TROOT::GetIconPath() {
+#ifdef ROOTICONPATH
+   if (IgnorePrefix()) {
+#endif
+      static TString rooticonpath;
+      if (rooticonpath.IsNull()) {
+         rooticonpath = "icons";
+         gSystem->PrependPathName(GetRootSys(), rooticonpath);
+      }
+      return rooticonpath;
+#ifdef ROOTICONPATH
+   } else {
+      const static TString rooticonpath = ROOTICONPATH;
+      return rooticonpath;
+   }
+#endif
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Get the fonts directory in the installation. Static utility function.
+
+const TString& TROOT::GetTTFFontDir() {
+#ifdef TTFFONTDIR
+   if (IgnorePrefix()) {
+#endif
+      static TString ttffontdir;
+      if (ttffontdir.IsNull()) {
+         ttffontdir = "fonts";
+         gSystem->PrependPathName(GetRootSys(), ttffontdir);
+      }
+      return ttffontdir;
+#ifdef TTFFONTDIR
+   } else {
+      const static TString ttffontdir = TTFFONTDIR;
+      return ttffontdir;
+   }
```

Comment:
Can we combine those 12 into (or a subset, e.g. the ROOTSYS-derived ones) into one, instead of having near-duplicate bodies? And yes, this might involve preprocessor macros :-(